File: hadron-desktop/src-tauri/src/commands/release_notes.rs
//! Release Notes Generator commands

use crate::database::Database;
use crate::jira_service;
use crate::release_notes_service::{self, ReleaseNotesConfig};
use std::sync::Arc;
use tauri::{AppHandle, State};

type DbState<'a> = State<'a, Arc<Database>>;

/// Full pipeline: fetch → enrich → generate → save
#[tauri::command]
pub async fn generate_release_notes(
    config: ReleaseNotesConfig,
    request_id: Option<String>,
    base_url: String,
    email: String,
    api_token: String,
    api_key: String,
    model: String,
    provider: String,
    db: DbState<'_>,
    app: AppHandle,
) -> Result<release_notes_service::ReleaseNotesResult, String> {
    log::debug!("cmd: generate_release_notes");
    log::info!(
        "Generating release notes for version {}",
        config.fix_version
    );
    release_notes_service::run_full_pipeline(
        config,
        &base_url,
        &email,
        &api_token,
        &api_key,
        &model,
        &provider,
        &db,
        &app,
        request_id.as_deref(),
    )
    .await
}

/// Dry-run: fetch and display tickets before generating
#[tauri::command]
pub async fn preview_release_notes_tickets(
    config: ReleaseNotesConfig,
    base_url: String,
    email: String,
    api_token: String,
) -> Result<Vec<release_notes_service::ReleaseNoteTicket>, String> {
    log::debug!("cmd: preview_release_notes_tickets");
    log::info!(
        "Previewing tickets for version {}",
        config.fix_version
    );
    release_notes_service::fetch_tickets_for_release(&config, &base_url, &email, &api_token).await
}

/// Populate fix version picker
#[tauri::command]
pub async fn list_jira_fix_versions(
    base_url: String,
    email: String,
    api_token: String,
    project_key: String,
) -> Result<Vec<jira_service::JiraFixVersion>, String> {
    log::debug!("cmd: list_jira_fix_versions");
    log::info!("Listing fix versions for {}", project_key);
    jira_service::list_fix_versions(base_url, email, api_token, project_key).await
}

/// Load a saved draft
#[tauri::command]
pub async fn get_release_notes(
    id: i64,
    db: DbState<'_>,
) -> Result<Option<crate::database::ReleaseNotesDraft>, String> {
    log::debug!("cmd: get_release_notes");
    let db = db.inner().clone();
    tokio::task::spawn_blocking(move || db.get_release_notes(id))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Database error: {}", e))
}

/// List drafts with pagination
#[tauri::command]
pub async fn list_release_notes(
    status: Option<String>,
    limit: Option<i64>,
    offset: Option<i64>,
    db: DbState<'_>,
) -> Result<Vec<crate::database::ReleaseNotesSummary>, String> {
    log::debug!("cmd: list_release_notes");
    let db = db.inner().clone();
    let limit = limit.unwrap_or(50);
    let offset = offset.unwrap_or(0);
    tokio::task::spawn_blocking(move || {
        db.list_release_notes(status.as_deref(), limit, offset)
    })
    .await
    .map_err(|e| format!("Task join error: {}", e))?
    .map_err(|e| format!("Database error: {}", e))
}

/// Save human edits
#[tauri::command]
pub async fn update_release_notes_content(
    id: i64,
    content: String,
    db: DbState<'_>,
) -> Result<(), String> {
    log::debug!("cmd: update_release_notes_content");
    let db = db.inner().clone();
    tokio::task::spawn_blocking(move || db.update_release_notes_content(id, &content))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Database error: {}", e))
}

/// Move through review workflow
#[tauri::command]
pub async fn update_release_notes_status(
    id: i64,
    status: String,
    reviewed_by: Option<String>,
    db: DbState<'_>,
) -> Result<(), String> {
    log::debug!("cmd: update_release_notes_status");
    let db = db.inner().clone();
    tokio::task::spawn_blocking(move || {
        db.update_release_notes_status(id, &status, reviewed_by.as_deref())
    })
    .await
    .map_err(|e| format!("Task join error: {}", e))?
    .map_err(|e| format!("Database error: {}", e))
}

/// Save checklist state
#[tauri::command]
pub async fn update_release_notes_checklist(
    id: i64,
    checklist_json: String,
    db: DbState<'_>,
) -> Result<(), String> {
    log::debug!("cmd: update_release_notes_checklist");
    let db = db.inner().clone();
    tokio::task::spawn_blocking(move || db.update_release_notes_checklist(id, &checklist_json))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Database error: {}", e))
}

/// Incremental update with new tickets
#[tauri::command]
pub async fn append_to_release_notes(
    id: i64,
    config: ReleaseNotesConfig,
    request_id: Option<String>,
    base_url: String,
    email: String,
    api_token: String,
    api_key: String,
    model: String,
    provider: String,
    db: DbState<'_>,
    app: AppHandle,
) -> Result<release_notes_service::ReleaseNotesResult, String> {
    log::debug!("cmd: append_to_release_notes");
    let db_inner = db.inner().clone();

    // Load existing draft
    let existing = db_inner
        .get_release_notes(id)
        .map_err(|e| format!("Database error: {}", e))?
        .ok_or_else(|| "Release notes not found".to_string())?;

    let existing_keys: Vec<String> = serde_json::from_str(&existing.ticket_keys)
        .unwrap_or_default();

    // Fetch new tickets
    release_notes_service::emit_progress_with_request(
        &app,
        release_notes_service::ReleaseNotesPhase::FetchingTickets,
        5.0,
        "Fetching new tickets...",
        request_id.as_deref(),
    );
    let all_tickets =
        release_notes_service::fetch_tickets_for_release(&config, &base_url, &email, &api_token)
            .await?;

    // Filter to only new tickets
    let new_tickets: Vec<_> = all_tickets
        .into_iter()
        .filter(|t| !existing_keys.contains(&t.key))
        .collect();

    if new_tickets.is_empty() {
        return Err("No new tickets found to append.".to_string());
    }

    // Generate incremental content
    let (combined, tokens, cost) = release_notes_service::apply_incremental_update(
        &existing.markdown_content,
        &new_tickets,
        &existing_keys,
        &config,
        &api_key,
        &model,
        &provider,
    )
    .await?;

    let mut all_keys = existing_keys.clone();
    all_keys.extend(new_tickets.iter().map(|t| t.key.clone()));
    let ticket_keys_json =
        serde_json::to_string(&all_keys).map_err(|e| format!("Serialization error: {}", e))?;
    let refreshed_tickets =
        release_notes_service::fetch_tickets_for_release(&config, &base_url, &email, &api_token)
            .await?;
    let insights = release_notes_service::compute_ai_insights(&refreshed_tickets, &combined);
    let insights_json = serde_json::to_string(&insights).ok();

    // Persist updated content and append metadata.
    db_inner
        .update_release_notes_after_append(
            id,
            &combined,
            &ticket_keys_json,
            all_keys.len() as i32,
            tokens,
            cost,
            insights_json.as_deref(),
        )
        .map_err(|e| format!("Database error: {}", e))?;

    release_notes_service::emit_progress_with_request(
        &app,
        release_notes_service::ReleaseNotesPhase::Complete,
        100.0,
        &format!("Appended {} new tickets", new_tickets.len()),
        request_id.as_deref(),
    );

    Ok(release_notes_service::ReleaseNotesResult {
        id,
        title: existing.title,
        markdown_content: combined,
        ticket_count: all_keys.len() as i32,
        ticket_keys: all_keys,
        ai_insights: Some(insights),
        tokens_used: tokens,
        cost,
        generation_duration_ms: 0,
    })
}

/// Export as markdown/confluence/html
#[tauri::command]
pub async fn export_release_notes(
    id: i64,
    format: String,
    db: DbState<'_>,
) -> Result<String, String> {
    log::debug!("cmd: export_release_notes");
    let db_inner = db.inner().clone();

    let draft = db_inner
        .get_release_notes(id)
        .map_err(|e| format!("Database error: {}", e))?
        .ok_or_else(|| "Release notes not found".to_string())?;

    match format.as_str() {
        "markdown" => Ok(draft.markdown_content),
        "confluence" => Ok(release_notes_service::markdown_to_confluence(
            &draft.markdown_content,
        )),
        "html" => Ok(release_notes_service::markdown_to_html(
            &draft.markdown_content,
        )),
        _ => Err(format!("Unsupported export format: {}", format)),
    }
}

/// Soft delete
#[tauri::command]
pub async fn delete_release_notes(
    id: i64,
    db: DbState<'_>,
) -> Result<(), String> {
    log::debug!("cmd: delete_release_notes");
    let db = db.inner().clone();
    tokio::task::spawn_blocking(move || db.soft_delete_release_notes(id))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Database error: {}", e))
}

/// On-demand style compliance check
#[tauri::command]
pub async fn check_release_notes_compliance(
    content: String,
    api_key: String,
    model: String,
    provider: String,
) -> Result<release_notes_service::ComplianceReport, String> {
    log::debug!("cmd: check_release_notes_compliance");
    log::info!("Running release notes compliance check");
    release_notes_service::check_compliance(&content, &api_key, &model, &provider).await
}

File: hadron-electron/src/services/release-notes.ts
/**
 * Release Notes Generator Service
 * Handles Tauri command invocations and config management for release notes.
 */

import { invoke } from "@tauri-apps/api/core";
import { getSetting, getApiKey } from "./secure-storage";
import { getStoredProvider, getStoredModel } from "./api";
import logger from "./logger";
import type {
  ReleaseNotesConfig,
  ReleaseNotesDraft,
  ReleaseNotesSummary,
  ReleaseNoteTicketPreview,
  JiraFixVersion,
  ReleaseNotesExportFormat,
  AiEnrichmentConfig,
  ReleaseNotesContentType,
  ComplianceReport,
} from "../types";

// ============================================================================
// JIRA Config Helpers
// ============================================================================

interface JiraCredentials {
  baseUrl: string;
  email: string;
  apiToken: string;
  projectKey: string;
}

async function getJiraCredentials(): Promise<JiraCredentials> {
  const baseUrl = (await getSetting<string>("jira_base_url", "")) || "";
  const email = (await getSetting<string>("jira_email", "")) || "";
  const apiToken = (await getApiKey("jira")) || "";
  const projectKey = (await getSetting<string>("jira_project_key", "")) || "";

  if (!baseUrl || !email || !apiToken) {
    throw new Error("JIRA is not configured. Please set up JIRA in Settings.");
  }

  return { baseUrl, email, apiToken, projectKey };
}

async function getAiCredentials(): Promise<{ apiKey: string; model: string; provider: string }> {
  const provider = getStoredProvider();
  const model = getStoredModel();
  const apiKey = await getApiKey(provider);

  if (!apiKey) {
    throw new Error("AI provider is not configured. Please set an API key in Settings.");
  }

  return { apiKey, model, provider };
}

// ============================================================================
// Fix Version Operations
// ============================================================================

export async function listFixVersions(projectKey?: string): Promise<JiraFixVersion[]> {
  try {
    const jira = await getJiraCredentials();
    const key = projectKey || jira.projectKey;

    if (!key) {
      throw new Error("No JIRA project key configured.");
    }

    const versions = await invoke<JiraFixVersion[]>("list_jira_fix_versions", {
      baseUrl: jira.baseUrl,
      email: jira.email,
      apiToken: jira.apiToken,
      projectKey: key,
    });

    logger.info("Listed fix versions", { count: versions.length, projectKey: key });
    return versions;
  } catch (error) {
    logger.error("Failed to list fix versions", { error });
    throw error;
  }
}

// ============================================================================
// Ticket Preview
// ============================================================================

export async function previewTickets(
  fixVersion: string,
  contentType: ReleaseNotesContentType = "both",
  jqlFilter?: string,
  moduleFilter?: string[],
): Promise<ReleaseNoteTicketPreview[]> {
  try {
    const jira = await getJiraCredentials();

    const config: ReleaseNotesConfig = {
      fixVersion,
      contentType,
      projectKey: jira.projectKey || undefined,
      jqlFilter,
      moduleFilter,
      aiEnrichment: {
        rewriteDescriptions: false,
        generateKeywords: false,
        classifyModules: false,
        detectBreakingChanges: false,
      },
    };

    const tickets = await invoke<ReleaseNoteTicketPreview[]>("preview_release_notes_tickets", {
      config,
      baseUrl: jira.baseUrl,
      email: jira.email,
      apiToken: jira.apiToken,
    });

    logger.info("Previewed tickets", { count: tickets.length, fixVersion });
    return tickets;
  } catch (error) {
    logger.error("Failed to preview tickets", { error });
    throw error;
  }
}

// ============================================================================
// Generation
// ============================================================================

export interface GenerateOptions {
  fixVersion: string;
  contentType: ReleaseNotesContentType;
  requestId?: string;
  jqlFilter?: string;
  moduleFilter?: string[];
  aiEnrichment?: Partial<AiEnrichmentConfig>;
}

export async function generateReleaseNotes(options: GenerateOptions) {
  try {
    const jira = await getJiraCredentials();
    const ai = await getAiCredentials();

    const config: ReleaseNotesConfig = {
      fixVersion: options.fixVersion,
      contentType: options.contentType,
      projectKey: jira.projectKey || undefined,
      jqlFilter: options.jqlFilter,
      moduleFilter: options.moduleFilter,
      aiEnrichment: {
        rewriteDescriptions: true,
        generateKeywords: true,
        classifyModules: true,
        detectBreakingChanges: true,
        ...options.aiEnrichment,
      },
    };

    const result = await invoke("generate_release_notes", {
      config,
      requestId: options.requestId || null,
      baseUrl: jira.baseUrl,
      email: jira.email,
      apiToken: jira.apiToken,
      apiKey: ai.apiKey,
      model: ai.model,
      provider: ai.provider,
    });

    logger.info("Generated release notes", { fixVersion: options.fixVersion });
    return result;
  } catch (error) {
    logger.error("Failed to generate release notes", { error });
    throw error;
  }
}

// ============================================================================
// CRUD Operations
// ============================================================================

export async function getReleaseNotes(id: number): Promise<ReleaseNotesDraft | null> {
  try {
    return await invoke<ReleaseNotesDraft | null>("get_release_notes", { id });
  } catch (error) {
    logger.error("Failed to get release notes", { id, error });
    throw error;
  }
}

export async function listReleaseNotes(
  status?: string,
  limit = 50,
  offset = 0,
): Promise<ReleaseNotesSummary[]> {
  try {
    return await invoke<ReleaseNotesSummary[]>("list_release_notes", {
      status: status || null,
      limit,
      offset,
    });
  } catch (error) {
    logger.error("Failed to list release notes", { error });
    throw error;
  }
}

export async function updateContent(id: number, content: string): Promise<void> {
  try {
    await invoke("update_release_notes_content", { id, content });
    logger.info("Updated release notes content", { id });
  } catch (error) {
    logger.error("Failed to update content", { id, error });
    throw error;
  }
}

export async function updateStatus(
  id: number,
  status: string,
  reviewedBy?: string,
): Promise<void> {
  try {
    await invoke("update_release_notes_status", {
      id,
      status,
      reviewedBy: reviewedBy || null,
    });
    logger.info("Updated release notes status", { id, status });
  } catch (error) {
    logger.error("Failed to update status", { id, error });
    throw error;
  }
}

export async function updateChecklist(id: number, checklistJson: string): Promise<void> {
  try {
    await invoke("update_release_notes_checklist", { id, checklistJson });
  } catch (error) {
    logger.error("Failed to update checklist", { id, error });
    throw error;
  }
}

// ============================================================================
// Incremental Update
// ============================================================================

export async function appendToReleaseNotes(
  id: number,
  fixVersion: string,
  contentType: ReleaseNotesContentType = "both",
  jqlFilter?: string,
  requestId?: string,
) {
  try {
    const jira = await getJiraCredentials();
    const ai = await getAiCredentials();

    const config: ReleaseNotesConfig = {
      fixVersion,
      contentType,
      projectKey: jira.projectKey || undefined,
      jqlFilter,
      aiEnrichment: {
        rewriteDescriptions: true,
        generateKeywords: true,
        classifyModules: true,
        detectBreakingChanges: true,
      },
    };

    const result = await invoke("append_to_release_notes", {
      id,
      config,
      requestId: requestId || null,
      baseUrl: jira.baseUrl,
      email: jira.email,
      apiToken: jira.apiToken,
      apiKey: ai.apiKey,
      model: ai.model,
      provider: ai.provider,
    });

    logger.info("Appended to release notes", { id });
    return result;
  } catch (error) {
    logger.error("Failed to append to release notes", { id, error });
    throw error;
  }
}

// ============================================================================
// Export
// ============================================================================

export async function exportReleaseNotes(
  id: number,
  format: ReleaseNotesExportFormat,
): Promise<string> {
  try {
    const content = await invoke<string>("export_release_notes", { id, format });
    logger.info("Exported release notes", { id, format });
    return content;
  } catch (error) {
    logger.error("Failed to export release notes", { id, format, error });
    throw error;
  }
}

// ============================================================================
// Delete
// ============================================================================

export async function deleteReleaseNotes(id: number): Promise<void> {
  try {
    await invoke("delete_release_notes", { id });
    logger.info("Deleted release notes", { id });
  } catch (error) {
    logger.error("Failed to delete release notes", { id, error });
    throw error;
  }
}

// ============================================================================
// Style Compliance
// ============================================================================

export async function checkCompliance(content: string): Promise<ComplianceReport> {
  try {
    const ai = await getAiCredentials();
    return await invoke<ComplianceReport>("check_release_notes_compliance", {
      content,
      apiKey: ai.apiKey,
      model: ai.model,
      provider: ai.provider,
    });
  } catch (error) {
    logger.error("Compliance check failed", { error });
    throw error;
  }
}
